DOC: Revise glossary page

bjnath opened the pull request:
Expanded and updated the Glossary page in an effort to make it more useful. 

This is a starting place; I welcome the community's comments regarding clarity, choice of entries, length of definitions, technical accuracy, etc.

bjnath committed:
DOC: Revise glossary page

Entries added, explanations revised,
Python-only entries now point to Python glossary.

bjnath committed:
DOC: Wording improvement in #16996

More clearly describe `shares_memory` as being
not always *feasible* rather than *possible*

BvB93 commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,552 @@
-.. automodule:: numpy.doc.glossary
+    (`n`,)
+
+      A tuple with one element.The trailing comma distinguishes a one-element
```
```suggestion
      A tuple with one element. The trailing comma distinguishes a one-element
```

BvB93 commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,552 @@
+
+    array_like
+
+      Any `sequence <https://docs.python.org/3/glossary.html#term-sequence>`_
```
```suggestion
      Any scalar or `sequence <https://docs.python.org/3/glossary.html#term-sequence>`_
```
I'd suggest mentioning the "scalar" part right from the get-go.

BvB93 commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,552 @@
+
+    .base
+
+      If an array does not own its memory, then its ``.base`` attribute
```
```suggestion
      If an array does not own its memory, then its :doc:`.base <reference/generated/numpy.ndarray.base>` attribute
```
Might be worthwhile to link to the ``ndarray.base`` method here, 
similar to how ``array()`` was referenced in the ``array_like`` section.

BvB93 commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,552 @@
+
+    field
+
+       In a :term:`structured data type`, each subtype is called a `field`.
```
```suggestion
       In a :term:`structured data type`, each subtype is called a :doc:`field <reference/generated/numpy.dtype.fields>`.
```

BvB93 commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,552 @@
+
+    shape
+
+      A tuple showing the length of each dimension of an ndarray. The
```
Might be an idea to link to ``generated/numpy.ndarray.shape`` somewhere.

BvB93 commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,552 @@
+
+    stride
+
+      Tuple of bytes to step in each dimension when traversing an array.
```
Same as above, maybe add a link to ``generated/numpy.ndarray.strides``?

BvB93 commented on doc/source/glossary.rst:
```suggestion
      needs to be physically distinct -- a :term:`copy`.
```

bjnath committed:
DOC: Incorporate @BvB93's suggestions in PR #16996

Also reworded `stride` entry for clarity, moved
'axis' entry to correct alphabetical position.

Also added anchors to reference/arrays.indexing for glossary references
to point to.

bjnath commented:
Thank you, @BvB93, excellent suggestions. All incorporated as-is except for the last (https://github.com/numpy/numpy/pull/16996#discussion_r465024333) -- if the user clicks the `copy` link, there's only "See views."

Also clarified `stride` entry, fixed alphabetization of `axis`, and replaced two URLs with intersphinx links.

BvB93 commented:
> All suggestions incorporated as-is except for the last (https://github.com/numpy/numpy/pull/16996#discussion_r465024333) -- if the user clicks the `copy` link, there's only "See views."

Huh, odd. Does the same issue apply to the ``:term:`view` `` references used earlier on in the document?

BvB93 commented:
Btw, there appear to be a couple of doctest-related issues ([link](https://app.circleci.com/pipelines/github/numpy/numpy/2889/workflows/2042ca6a-c02b-4b0d-861a-52d86bde942e/jobs/14979/parallel-runs/0/steps/0-105)).
I suspect that these errors are preventing circle from rendering the updated [docs](https://14979-908607-gh.circle-artifacts.com/0/doc/build/html/index.html).

bjnath committed:
DOC: Avoid unwanted doctest on PR #14979

Failing CI because it's running doctest on the example strings.

Also, deleting glossary.py, source file of the original glossary.

bjnath committed:
DOC: 2nd try: Avoid unwanted doctest on PR #14979

bjnath commented:
@BvB93 
> if the user clicks the copy link, there's only "See views."

Sorry, I wasn't clear. The glossary handles both `copy` and `view` in the `view` entry and the `copy` entry just says 'see view'.

As a result, there's no value in giving `copy` a `:term:` role inside the `view` entry, since the link points to the `copy` entry which just says to look at the `view` entry.

bjnath committed:
DOC: 3rd try: Avoid unwanted doctest on PR #14979

bjnath committed:
DOC: 4th try: Avoid unwanted doctest on PR #14979

bjnath committed:
DOC: 5th try: Avoid unwanted doctest on PR #14979

BvB93 commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,542 @@
+
+      It can be used at most once:
+
+        >>> a[0,...,0,...].shape
```
This should do the trick:
```python
>>> a[0,...,0,...].shape
Traceback (most recent call last):
  ...
IndexError: an index can only have a single ellipsis ('...')
```

bjnath replied:
I threw up my hands in the latest commit, but this is good to know, thanks.

BvB93 replied:
For future reference: the official `doctest` documentation goes into a bit more detail about exceptions and doctesting ([docs](https://docs.python.org/3/library/doctest.html#what-about-exceptions)).

bjnath replied:
I actually was on that page. I read "just paste in the traceback" in the first sentence and never reached the truthful part.  Gotta love that developer swagger. Now that I know, I'll have a serious look. Thanks.

bjnath committed:
DOC: 6th try: Fix doctest on PR #16996

bjnath commented:
We present at last https://15004-908607-gh.circle-artifacts.com/0/doc/build/html/glossary.html

WarrenWeckesser commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,532 @@
+
+    subarray data type
+
+      An element of a strctured datatype that behaves like an ndarray.
```
```suggestion
      An element of a structured datatype that behaves like an ndarray.
```

bjnath replied:
Thanks -- fixed.

bjnath replied:
Fixed, thanks.

WarrenWeckesser commented:
Nice!

bjnath committed:
DOC: Fix typos in #16996

strctured -> structured, caught by @WarrenWeckesser.

Fixed missing language in `view` entry.

BvB93 commented:
References to the old `doc.glossary` module will also have to be removed from certain tests ([ref](https://github.com/bjnath/numpy-1/blob/780ea6002be4bd7e1d07287fba88eac4ff45157e/numpy/tests/test_public_api.py#L154)).

Not entirely sure though if we're allowed to straight up remove the old `doc.glossary` module; it is apparently part of the public API. 
A second opinion would be very much appreciated here.

bjnath commented:
@WarrenWeckesser  -- Thank you!

bjnath committed:
DOC: Fix repeated word in PR #16996

Also adding <BLANKLINE>s to some examples for clarity.

seberg commented:
I think we may be able to pull removal of, maybe discuss it today.  But for example debian strips the whole `doc` module to begin with and nobody seems to have ever complained.  And I guess that `np.info` (which can pick these up) is rarely to never used.

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -2,6 +2,533 @@
+    array_like
+
+      Any :doc:`scalar <reference/arrays.scalars>` or
+      `sequence <https://docs.python.org/3/glossary.html#term-sequence>`_
```
Can be done in a follow-up, but would be nice to replace these with intersphinx links.

bjnath replied:
I wondered if there weren't a better way to handle these. I'm glad to fix it right away. What should it say?

eric-wieser replied:
My guess would be ``:term:`python:sequence` ``, but that's only a guess.

bjnath replied:
This is a working link, because we have an intersphinx link in `conf.py`:
```
      :ref:`slice <python:glossary>`
```
but we want to link to the `slice` entry, not just the page. This gives "undefined label":
```
      :ref:`slice <python:glossary#term-slice>`
```
as does this:
```
      :ref:`slice <python:glossary:term:slice>`
```
When it comes to half-assed, Sphinx is without peer. 

Leaving the URLs for now.

eric-wieser replied:
Did my approach fail?

This seems to suggest it should work: https://stackoverflow.com/q/46473481/102441

rossbar replied:
I went ahead and swapped all the explicit links using intersphinx in c24bd and 690037.

I'm against duplicating the terms from the Python glossary in NumPy's glossary, as it just leads to more clicks for the user. Nevertheless I switched them to intersphinx links in 6900372 and leave the decision whether to include them to others.

@bjnath a helpful resource for working with intersphinx is to inspect the objects.inv file, which describes the mapping from the sphinx names to URLs. It's not the most visible thing in the world, but [the intersphinx docs](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#showing-all-links-of-an-intersphinx-mapping-file) explain how to grab `objects.inv` for inspection.

rossbar replied:
> Did my approach fail?

The approach works as expected! I went ahead and applied it throughout.

bjnath replied:
Think I stole the half-assed prize. Sorry @eric-wieser, had the answer right in front of me. Thanks @rossbar for the fixes.

bjnath replied:
Also @rossbar thanks for the `objects.inv` tip.

bjnath replied:
I'll write a guide to linking for NumPy docs writers, including stuff like neps and of course this one.

bjnath committed:
DOC: Fix example formatting in PR #16996

rossbar committed:
DOC: Use intersphinx links.

rossbar committed:
DOC: intersphinx standalone glossary entries

rossbar committed:
DOC: rm tuple from glossary

bjnath commented:
@rossbar:

> I'm against duplicating the terms from the Python glossary in NumPy's glossary, as it just leads to more clicks for the user

Linking to the Python glossary alerts you that the unfamiliar term isn't NumPy, it's Python, and it takes you directly to the definition.

It's easy to imagine technical users who are new to both Python and NumPy. It's a good deed and good documentation to help them sort it out. We can do it adequately with a small number of links.

Neither of these sounds better:

* Copy the Python terms into our glossary

* Leave the terms out and provide no clue where to look

Also, could you help me understand this change?

> DOC: rm tuple from glossary 

I kept many of the Python terms from the original glossary and perhaps we don't need them all, though it's not clear what harm they do.

Why the beef with `tuple`? It's pervasive in NumPy and seems worth setting straight for anyone who thinks they know what it is but isn't sure.

rossbar commented:
> Neither of these sounds better:

Agreed - I would prefer that we use intersphinx for the terms where they appear. In other words, if in the documentation it says e.g. `ndarray has the attribute shape`, instead of linking to our glossary via ``ndarray has the :term:`attribute` shape``, we do ``ndarray has the :term:`python:attribute` shape``. The reason being that the former takes you to a link in the numpy glossary, which is itself a link to the Python glossary, requiring multiple clicks to get to the definition. I think it would be better if it were a single click. This should be an easy find/replace for the duplicate terms.

> Also, could you help me understand this change?

It turns out there is no glossary entry for "tuple" in the Python glossary!

bjnath commented:
> Agreed - I would prefer that we use intersphinx for the terms where they appear. In other words, if in the documentation it says e.g. ndarray has the attribute shape, instead of linking to our glossary via ndarray has the :term:`attribute` shape, we do ndarray has the :term:`python:attribute` shape

Thanks for explaining this. I'd intended for them to work this way. I will fix it.  But we can still keep the Python links to help people who encounter those terms in other NumPy docs. 

>  no glossary entry for "tuple"

Two "d'ohs" in a single PR! Thanks for catching this. I did try to check the links but need to do it more effectively.

bjnath commented:
@rossbar -- So to be clear, your wish is that when a non-glossary file has a Python :term:, it should link directly to the Python glossary rather than to ours? That would be my preference too.

A quick grep for :term: doesn't seem to turn up many, or possibly any.

Actually most of the :term: links seem silly; they're from files that themselves define the terms; visiting the glossary won't turn up anything that hasn't already been explained in the file. I'd like to fix these, but in another PR.

Since there don't seem to be many long-way-around Python links, and the existing Python links might serve a useful purpose, are you OK on leaving the Python links in the glossary?

rossbar commented:
> So to be clear, your wish is that when a non-glossary file has a Python :term:, it should link directly to the Python glossary rather than to ours?

Yes, linking directly to the relevant glossary would be preferable.

> Since there don't seem to be many long-way-around Python links, and the existing Python links might serve a useful purpose, are you OK on leaving the Python links in the glossary?

I don't feel strongly about it - my main concern was with the multiple-redirection problem. If we feel it provides useful context to have stubs in the NumPy glossary to point to the Python glossary then that's fine by me!

mattip commented:
I think we are ok with this change which will clear the test failures
```
--- a/numpy/tests/test_public_api.py
+++ b/numpy/tests/test_public_api.py
@@ -151,7 +151,6 @@ PUBLIC_MODULES = ['numpy.' + s for s in [
     "doc.constants",
     "doc.creation",
     "doc.dispatch",
-    "doc.glossary",
     "doc.indexing",
     "doc.internals",
     "doc.misc",

```

mattip commented:
CI is passing. I think this is ready to merge.

rossbar commented:
A lot of the definitions have changed significantly, and there are many new definitions that should probably be carefully reviewed. I started to do so but the number of comments/questions was ballooning quite a bit so I stopped.

If we'd like to get the structural changes in and worry about reviewing the new/changed definitions in a follow-up PR to prevent this one from having too many comments/questions that's fine with me.

mattip commented:
I think we should do the opposite: move the file as-is from `numpy/doc/glossary.py` to `doc/source/glossary.rst` (and make CI happy) in one PR, and then update the definitions in subsequent work.

mattip commented:
It might be helpful, if possible, to break the changes into smaller groups of related items for review.

bjnath commented:
> break the changes into smaller groups

That's a very good idea.

eric-wieser commented:
Agreed with @mattip that we should really handle moves and improvements as separate PRs, as reviewing them simultaneously is very difficult.

bjnath commented:
I've already agreed to both @mattip's suggestions --  
1. move the contents of the py original to this rst
2. resubmit glossary updates piecemeal in smaller PRs.

My intention was only to update the glossary. We stumbled over the doc.glossary skeleton en route.  I'm all for a separation of concerns that buries the skeleton before attempting a doc review.

mattip commented:
I rebased this to make comparisons easier after gh-17222, and created companion PRs to remove some terms and add new ones. Once those PRs are resolved, we can rebase once again to continue to work on the modications here (or someone could add another PR with the only the changes).

mattip committed:
adjust whitespace to minimize the diff

mattip committed:
DOC: manually merge some changes done in the other PRs

mattip commented on doc/source/glossary.rst:
```diff
@@ -321,7 +327,10 @@ Glossary
 
 
    `little-endian <https://en.wikipedia.org/wiki/Endianness>`_
-       \
```
I wasn't sure for little-endian and big-endian if we decided to leave the term as a link to wikipedia or not, so I left this one and changed the other.

mattip commented on doc/source/glossary.rst:
```diff
@@ -390,7 +394,20 @@ Glossary
+
+       Row-major order is also known as the C order, as the C programming
+       language uses it. New NumPy arrays are by default in row-major order.
+
```
Was the content here intentionally removed in this PR?

mattip commented on doc/source/glossary.rst:
```diff
@@ -458,10 +475,6 @@ Glossary
-   type
-       In NumPy, a synonym for :term:`dtype`. For the more general Python
-       meaning, :term:`see here. <python:type>`
-
```
This entry has dissapeared on master, so I removed it from the PR as well. OK?

bjnath replied:
This was new content, not meant to be removed.

mattip committed:
Merge branch 'master' into revised-glossary

mattip commented:
Merged the master HEAD into this branch and resolved the differences.

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -95,54 +95,55 @@ Glossary
-       Many operations can take place along one of these axes.  For example,
-       we can sum each row of an array, in which case we operate along
-       columns, or axis 1::
+       The result of an operation along an :term:`axis` X is an array in which X
+       disappears. This can surprise new users expecting the opposite.
```
This description is inaccurate:
```
>>> x = np.arange(12).reshape(3, 4)
>>> x.shape
(3, 4)
>>> np.sort(x, axis=0).shape  # sort along axis 0
(3, 4)  # axis didn't disappear!
```

It would be accurate as
```suggestion
       The result of an reduction along an :term:`axis` X is an array in which X
       disappears. This can surprise new users expecting the opposite.
```
but I suspect you want this glossary entry to describe the more general case.

bjnath replied:
Thanks for pointing this out. Alternative in 511d769 .

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -95,54 +95,55 @@ Glossary
+       :doc:`numpy.array <reference/generated/numpy.array>`
+       is array_like. ::
+
+           >>> a = np.array([[1, 2.0],[0, 0],(1+1j, 3.)])
```
PEP8
```suggestion
           >>> a = np.array([[1, 2.0], [0, 0], (1+1j, 3.)])
```

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -226,23 +226,21 @@ Glossary
-         >>> x + y
-         array([[4, 5],
-                [5, 6]])
+       Ordinarily this means the shapes of a and b must be identical. But in
```
```suggestion
       Ordinarily, this means the shapes of a and b must be identical. But in
```

bjnath replied:
[Sentence adverbs](https://www.merriam-webster.com/dictionary/sentence%20adverb) usually need the comma:
> Frankly, that's impossible.

but in this case it's stylistic:

> Ordinarily X is rational; on weekends it's irrational.
> Ordinarily, X is rational; on weekends, it's irrational.

and we can imagine a single text having both:

> We show that usually X is rational but on weekends it's irrational.
> Usually, X is rational, but, on weekends, it is irrational. QED.

eric-wieser replied:
Thanks for giving me the name of this construction.

Both https://www.grammarly.com/blog/commas-after-introductory-words/ and https://theeditorsblog.net/2016/02/21/a-tale-of-adverbs-and-the-comma/ suggest that the comma should be present, so I'd err towards matching the norm rather than flouting stylistic freedom.

bjnath replied:
I wouldn't define "norms" based on what's on the internet.

bjnath replied:
The [guidance given by the Grammarly program (pg88)](http://onlinewritingcenters.org/wp-content/uploads/2019/01/2017-Dembsey-WCJ-36.1.pdf) appears to be:
> Introductory clauses and long phrases must always be followed by a comma, while single words and short phrases may or may not.

The second reference doesn't seem to say anything about this case.

Neither one seems to be supporting the comma campaign.

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -257,67 +255,40 @@ Glossary
-         ...         return f(*args, **kwargs)
-         ...
-         ...     return new_logging_func
+   :term:`decorator <python:decorator>`
+       \
```
Does this work / makes sense in a glossary? Where does ``:term:`decorator` `` link to?

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -257,67 +255,40 @@ Glossary
-       the integer 1 to the string "one", and the string "two" to
-       the list ``[1, 2]``.  The values may be accessed using their
-       corresponding keys::
+       The datatype describing the (identically typed) elements in an ndarray.
+       It can be changed to reinterpret the array contents. For details, see
+       :doc:`Data type objects (dtype). <reference/arrays.dtypes>`
 
-         >>> x[1]
-         'one'
 
-         >>> x['two']
-         [1, 2]
+   fancy indexing
 
-       Note that dictionaries are not stored in any specific order.  Also,
-       most mutable (see *immutable* below) objects, such as lists, may not
-       be used as keys.
+       Another term for :term:`advanced indexing`.
```
Merging has gone wrong here, these entries are repeated below

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -344,122 +315,98 @@ Glossary
-         >>> x = [2, 2.0, "two", [2, 2.0]]
+   homogeneous
+       All elements of a homogeneous array have the same type. ndarrays, in
+       contrast to Python lists, are homogeneous. The type can be complicated,
+       as in a :term:`structured array`, but all elements have that type.
 
-       The list `x` contains 4 items, each which can be accessed individually::
+      NumPy `object arrays <#term-object-array>`_, which contain references to
+      Python objects, fill the role of heterogeneous arrays.
```
Indent is inconsistent here, and will likely render badly.

Also, should be ``:term:`object-array` ``

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -344,122 +315,98 @@ Glossary
 
-   ndarray
-       See *array*.
+   object array
```
Bad merge, this is repeated on line 412

eric-wieser commented on doc/source/glossary.rst:
```suggestion
       An array created this way is a "view", and the performance gain often
```
This isn't really a code snippet

bjnath replied:
[Google style](https://developers.google.com/style/key-terms) for this is italics, will change.

eric-wieser replied:
Works for me, I can just never remember the rst syntax

eric-wieser reviewed (commented):
Seems like the merge didn't go cleanly in a few places.

bjnath committed:
DOC: Updated for reviewer comments

Also added "contiguous," which is referenced in docs but hadn't been
defined.

bjnath commented:
Pushed update to address comments above. Also adds "contiguous," which had not been in the glossary but had a reference in the docs.

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -95,54 +99,62 @@ Glossary
+           array([[ 0,  1,  2,  3],
+                  [ 4,  5,  6,  7],
+                  [ 8,  9, 10, 11]])
+           <BLANKLINE>
```
This looks like a misuse of blankline - numpy does not emit a blank line here.

bjnath replied:
It might be misleading if the text were part of a docstring, but here it's used for formatting the documentation and is invisible to the reader.

bjnath replied:
Sorry, missed your point initially. Agreed, will change.

bjnath replied:
Removed in 511d769.

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -214,126 +225,58 @@ Glossary
-       sequence addresses/sends/stores the most significant byte first (lowest
-       address) and the least significant byte last (highest address). Common in
-       micro-processors and used for transmission of data over network protocols.
+   `big-endian <https://en.wikipedia.org/wiki/Endianness>`_
```
Are you sure these work as targets for `` :term:`big-endian` ``?

bjnath replied:
You can try it: Such a link appears in the > entry; it takes you to the big-endian glossary entry, which is itself a working link to the Wikipedia article; similarly for little.

eric-wieser replied:
You're right, this seems to work. As a reader of the HTML docs though, it's not at all obvious to me that the definitions are links, especially if they're already highlighted due to being the focused link target.

For that reason, I think a simple description of `see ...` might be clearer, especially since we use that elsewhere in the glossary.

bjnath replied:
Added in 563f5c27a.

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -344,199 +287,113 @@ Glossary
-         >>> x[mask] = -1
-         >>> x
-         array([ 0,  1,  2,  -1, -1])
+       The boolean array used to select elements in a :term:`masked array`.
```
I think you need to give both meanings here. The old meaning was much more frequently needed than this new one.

bjnath replied:
You're right. 511d769 restores old definition.

eric-wieser reviewed (commented):
Conflicts look resolved now, thanks.

bjnath commented on doc/source/glossary.rst:
It just shows as italic.

bjnath committed:
DOC: 'Along an axis' rewrite and other reviewer changes

bjnath commented on doc/source/glossary.rst:
```diff
@@ -37,8 +41,8 @@ Glossary
 
        It can be used at most once; ``a[...,0,...]`` raises an :exc:`IndexError`.
```
Thanks -- fixed in 511d76940.

bjnath commented on doc/source/glossary.rst:
```diff
@@ -95,54 +99,62 @@ Glossary
+                  [ 7,  6,  5,  4],
+                  [11, 10,  9,  8]])
 
+       An operation "along axis 0" behaves as if the argument were an array of
```
Good suggestion; rewritten in 511d76940.

bjnath commented on doc/source/glossary.rst:
```diff
@@ -214,126 +225,58 @@ Glossary
 
-           [1, 4, 2, 5, 3, 6]
+   contiguous
+       An array is contiguous if it occupies a single unbroken block of memory.
```
Thanks -- modified in 511d76940

bjnath commented on doc/source/glossary.rst:
```diff
@@ -555,56 +412,75 @@ Glossary
-       A :term:`structured data type` may contain a :term:`ndarray` with its
-       own dtype and shape:
+   subarray
+      An array nested in a :term:`structured data type`: ::
```
Change made in 511d76940.

bjnath commented on doc/source/glossary.rst:
I'm not sure what I was trying to convey. You're right. Changed in 511d76940.

bjnath commented on doc/source/glossary.rst:
```diff
@@ -14,12 +14,16 @@ Glossary
        Used as a dimension entry, ``-1`` instructs NumPy to choose the length
        that will keep the total number of elements the same.
 
+           >>> np.arange(12).reshape(4,-1).shape
```
Added in 511d76940.

I had great fun trying to get intersphinx to work on the python link. `conf.py` says
```
intersphinx_mapping = {
    'python': ('https://docs.python.org/dev', None),
```
and calling
```
 python -m sphinx.ext.intersphinx https://docs.python.org/dev/objects.inv
```
shows
```
std:doc
        faq/programming                          Programming FAQ                         : faq/programming.html
```
so you'd think
```
doc:`python:Programming FAQ`
```
would work. It didn't. Even if it did, it wouldn't link to the relevant entry.

Maybe I missed something. Happy to replace the URL with something Sphinx that works.

eric-wieser replied:
``doc:`python:faq/programming` ``?

eric-wieser replied:
I'm not seeing the Sphinx link you're referring to in this diff context.

bjnath replied:
Because I chose to keep the link that went directly to the entry, even though it was a URL.

eric-wieser replied:
Using the url looks like the right choice here, unless Sphinx has some hybrid "append this path to the base url of this intersphinx target". Avoiding urls is most important for internal links. For external ones like these, the only danger is linking to an outdated python doc version.

bjnath commented on doc/source/glossary.rst:
Missed this in 90111e2

bjnath replied:
Changed in 563f5c27a

bjnath committed:
DOC: Replace linked headings with explicit 'See'

bjnath commented:
Latest: https://15820-908607-gh.circle-artifacts.com/0/doc/build/html/glossary.html

bjnath committed:
DOC: Fix minor typos

bjnath commented:
Are we ready to take the final step and merge this?

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -39,13 +42,13 @@ Glossary
          large arrays. To see the entire array, use `numpy.printoptions`
 
 
-   ``:``
```
Any reason for dropping the code formatting?

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -99,35 +102,41 @@ Glossary
-       Reversing along axis 0 (the row axis) below reverses the columns, and
-       reversing along axis 1 (the column axis) reverses the rows. This might
-       be the opposite of what a new user would expect.
+       An operation `along axis n` of array ``a`` behaves as if its argument
```
Using backticks for emphasis probably isn't a good idea: numpydoc will try to match them against links wherever possible, using italic text only as a fallback.

Rst syntax for italics is `*text*`.

eric-wieser replied:
Ah nevermind, this refers to the glossary heading so is fine

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -259,7 +279,10 @@ Glossary
+       An array is contiguous if
+           * it occupies an unbroken block of memory, and
+           * array elements with higher indexes occupy higher addresses (that
+             is, the :term:`stride` is not negative).
```
```suggestion
             is, no :term:`stride` is negative).
```

bjnath replied:
Nice. Changed in 73e95dd.

bjnath commented:
@eric-wieser  
> Any reason for dropping the code formatting?

I'd added the styling for visibility, since the punctuation headings are hard to see. I dropped it since the new theme shrinks the code font, making the headings even harder to see.

bjnath committed:
DOC: Revise 'contiguous' in PR #16996

eric-wieser commented:
> I'd added the styling for visibility, since the punctuation headings are hard to see.

Perhaps "`:` (symbol)"?

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -206,337 +224,206 @@ Glossary
 
    broadcast
-       NumPy can do operations on arrays whose shapes are mismatched::
+       ``broadcasting`` is NumPy's ability to process ndarrays of
```
```suggestion
       "broadcasting" is NumPy's ability to process ndarrays of
```
or
```suggestion
       `broadcasting` is NumPy's ability to process ndarrays of
```

bjnath replied:
In 33b25bc

bjnath commented:
>> I'd added the styling for visibility, since the punctuation headings are hard to see.

> Perhaps ": (symbol)"?

Why not, gave it a shot. 

![image](https://user-images.githubusercontent.com/6691888/94344140-3cc28580-ffeb-11ea-8166-a2e7d3ef4bb5.png)

"(symbol)" overachieves the goal of catching the eye and swallows the symbol.  There's no patching up the fact that the monospaced font is too small.

I can't think of a compelling rationale for using one format or the other, so I'd like the freedom to pick whatever's easiest to see.

bjnath committed:
DOC: Italics sted monospace in PR #16996

eric-wieser commented:
Thanks for trying that out, I agree it looks better as is.

eric-wieser commented on doc/source/glossary.rst:
```diff
@@ -6,20 +6,27 @@ Glossary
 
-       **When indexing an array**, shorthand that the missing axes, if they
-       exist, are full slices.
+   . . .
```
Any reason for the spaces here? That's likely to frustrate users searching for `...` on this page.
```suggestion
   ...
```

bjnath replied:
Legibility, of course.  I thought it more likely that someone would eyeball the glossary for the ellipsis, especially since it's on top of the page, than make the effort of searching with the browser. 

Sphinx search doesn't find either version. 

Here's how it looks without spaces:

![image](https://user-images.githubusercontent.com/6691888/94345245-8fec0680-fff2-11ea-992a-9b7cd81734f3.png)

I think the spaces help more users than they hurt.

bjnath commented:
Do any issues remain, or can this be merged?

mattip commented:
Let's merge this as-is, and tweak it in subsequent PRs

mattip commented:
Thanks @bjnath

bjnath commented:
And thank you, @mattip, for untangling and re-threading what turned out to be a much more complicated PR than I thought. If you hadn't stepped in I'd still be looking at the pieces like a dazed Ikea purchaser. I appreciate your putting thought and effort into making this merge possible.

bjnath commented:
And thanks also to reviewers @BvB93 , @WarrenWeckesser, @rossbar, and @eric-wieser!
